crates/flux-infer/src/refine_tree.rs

```diff
@@ -780,3 +791,328 @@ impl RefineCtxtTrace {
         Self { bindings, exprs }
     }
 }
+
+/* ---------------------------------------------------------------------------------------
+ BOT elimination
+
+ Build a graph G = (V, E) where
+
+     V = KVar or CONC
+     E = { v1 ---> v2 if there is a constraint where v1 appears as ASSM and v2 as HEAD }
+
+ let Post*(CONC) = {k | CONC -->* k in G}
+
+ forall each k not-in Post*(CONC)
+     set k to FALSE;
+     delete constraints where k appears as ASSM
+--------------------------------------------------------------------------------------- */
+
+/* ---------------------------------------------------------------------------------------
+ TOP elimination
+
+ Build a graph G = (V, E) where
+
+     V = KVar or CONC
+     E = { v1 ---> v2 if there is a constraint where v1 appears as ASSM and v2 as HEAD }
+
+ let Pre*(CONC) = {k | k -->* CONC in G}
+
+ forall each k not-in Pre*(CONC)
+     set k to TRUE
+--------------------------------------------------------------------------------------- */
+
+impl Node {
+    /// replace bot-kvars with false
+    fn simplify_bot(&mut self, genv: GlobalEnv, assumed_preds: &mut SnapshotMap<Expr, ()>) {
+        let graph = Graph::new(self);
+        let bots = self.bot_kvars(graph);
+        self.simplify_with_assignment(&bots);
+        // prune
+        self.simplify(genv, assumed_preds);
+    }
+
+    /// replace top-kvars with true
+    fn simplify_top(&mut self, genv: GlobalEnv, assumed_preds: &mut SnapshotMap<Expr, ()>) {
+        let graph = Graph::new(self);
+        let tops = self.top_kvars(graph);
+        self.simplify_with_assignment(&tops);
+        // prune
+        self.simplify(genv, assumed_preds);
+    }
+
+    /// simplifies assm and head using the TOP/BOT kvar assignment; follow
+    /// with a call to `simplify` to delete constraints with FALSE assm.
+    fn simplify_with_assignment(&mut self, assignment: &Assignment) {
+        match &mut self.kind {
+            NodeKind::Head(pred, tag) => {
+                let pred = assignment.simplify(pred);
+                self.kind = NodeKind::Head(pred, *tag);
+            }
+            NodeKind::Assumption(pred) => {
+                let pred = assignment.simplify(pred);
+                self.kind = NodeKind::Assumption(pred);
+            }
+            _ => {}
+        }
+        for child in &self.children {
+            child.borrow_mut().simplify_with_assignment(assignment);
+        }
+    }
+
+    /// Computes the set of all kvars that can be assigned to Bot (False),
+    /// because they are not (transitively) reachable from any concrete ASSUMPTION.
+    fn bot_kvars(&self, graph: Graph) -> Assignment {
```

Comment:
This is not using `&self`, and it could be a method in `Graph`.

Reply:
oh yes, forgot to move them!